File: apple/LC186ReverseWordsInAStringII.java
class Solution {
    public void reverseWords(char[] str) {
        int start = 0;
        for (int i = 0; i <= str.length; i++) {
            if (i == str.length || str[i] == ' ') {
                tempswap(str, start, i-1);
                start = i+1;
            }
        }
        tempswap(str, 0, str.length-1);
    }

    public void tempswap(char[] str, int i, int j) {
        while (i < j) {
            char c = str[i];
            str[i++] = str[j];
            str[j--] = c;
        }
    }
}

File: EA/LC186ReverseWordsInAStringII.py
class Solution(object):
    # input is a char array, not a strict string
    def reverseWords(self, string):
        if not string:
            return
        string.append(' ')
        start = 0
        i = 0
        while i < len(string):
            if string[i] == ' ':
                self._swap(string, start, i-1)
                start = i+1
            i += 1
        string.pop()
        print len(string)
        self._swap(string, 0, len(string)-1)

    def _swap(self, string, start, end):
        while start < end:
            string[start], string[end] = string[end], string[start]
            start += 1
            end -= 1
